Doc/whatsnew/3.11.rst

```diff
@@ -493,6 +488,226 @@ Optimizations
+
+Python will only specialize
+when it sees code that is "hot" (executed multiple times). This prevents Python
+from wasting time for run-once code. Python can also de-specialize when code is
+too dynamic or when the use changes. Specialization is attempted periodically,
+and specialization attempts are not too expensive. This allows specialization
+to adapt to new circumstances.
+
+(PEP written by Mark Shannon, with ideas inspired by Stefan Brunthaler.
+See :pep:`659` for more information.)
+
+..
+   If I missed out anyone, please add them.
+
++---------------+--------------------+-------------------------------------------------------+-------------------+-------------------+
+| Operation     | Form               | Specialization                                        | Operation speedup | Contributor(s)    |
+|               |                    |                                                       | (up to)           |                   |
++===============+====================+=======================================================+===================+===================+
+| Binary        | ``x+x; x*x; x-x;`` | Binary add, multiply and subtract for common types    | 10%               | Mark Shannon,     |
+| operations    |                    | such as ``int``, ``float``, and ``str`` take custom   |                   | Dong-hee Na,      |
+|               |                    | fast paths for their underlying types.                |                   | Brandt Bucher,    |
+|               |                    |                                                       |                   | Dennis Sweeney    |
++---------------+--------------------+-------------------------------------------------------+-------------------+-------------------+
+| Subscript     | ``a[i]``           | Subscripting container types such as ``list``,        | 10-30%            | Irit Katriel,     |
+|               |                    | ``tuple`` and ``dict`` directly index the underlying  |                   | Mark Shannon      |
+|               |                    | data structures.                                      |                   |                   |
+|               |                    |                                                       |                   |                   |
+|               |                    | Subscripting custom ``__getitem__``                   |                   |                   |
+|               |                    | is also inlined similar to :ref:`inline-calls`.       |                   |                   |
++---------------+--------------------+-------------------------------------------------------+-------------------+-------------------+
+| Store         | ``a[i] = z``       | Similar to subscripting specialization above.         | 10-90%            | Dennis Sweeney    |
+| subscript     |                    |                                                       |                   |                   |
```

Comment:
So IIUC from your comment (so many comments! :-( ) the problem isn't that 90% seems high, it is that 30% is low?! I worry that we're claiming extreme improvements here that nobody will be able to repro in real life. Irit's benchmark for `__getitem__` seems to only indicate up to 25%.


And using a timeit microbenchmark similar to Irit's I can't show more than a modest speed improvement for subscript assignment:
```
PS C:\Users\gvanrossum\cpython> py -3.10 -m timeit -s "a = [0,1,2,3,4,5]" "a[0] = 0"
10000000 loops, best of 5: 22.9 nsec per loop

PS C:\Users\gvanrossum\cpython> .\installed\python.exe -m timeit -s "a = [0,1,2,3,4,5]" "a[0] = 0"
20000000 loops, best of 5: 18.9 nsec per loop
```

Reply:
Looking back, I think my benchmarks were like this, so an even finer level of micro-ness. I guess this is the tragedy of microbenchmarks.

```python
from pyperf import Runner
runner = Runner()

runner.timeit("dict[str]=...",
    setup="from itertools import repeat; d = dict()",
    stmt="for s in repeat('foobar', 100_000):\n"
         "    d[s] = d[s] = d[s] = d[s] = d[s] = d[s] = d[s] = d[s] = d[s] = d[s] = s"
)
runner.timeit("list[int]=...",
    setup="from itertools import repeat; L = [0]*10",
    stmt="for i in repeat(2, 100_000):\n"
         "    L[i] = L[i] = L[i] = L[i] = L[i] = L[i] = L[i] = L[i] = L[i] = L[i] = i"
)
```

Reply:
Yeah I think I based it off of Dennis' microbenchmarks which aren't compatible  fully with Irit's. I will reduce it to a more moderate number (25%).